AADServicePrincipal: AppRoleAssignedTo parameter referencing issue

### Description of the issue

DSC running into errors when referencing `AppRoleAssignedTo` parameter for [AADServicePrincipal](https://microsoft365dsc.com/resources/azure-ad/AADServicePrincipal/).

We are getting the below error:

```
##[error]Cannot bind argument to parameter 'ReferenceObject' because it is null.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.CompareObje 
   ctCommand
    + PSComputerName        : localhost
```

Upon further investigation it looks like the module has a bug at [this](https://github.com/microsoft/Microsoft365DSC/blob/447773b96aa57d642a96bf5871730db5115f2807/Modules/Microsoft365DSC/DSCResources/MSFT_AADServicePrincipal/MSFT_AADServicePrincipal.psm1#L609) location in the module.


### Microsoft 365 DSC Version

V1.25.122.2

### Which workloads are affected

other

### The DSC configuration

```powershell
AADServicePrincipal 'AADServicePrincipal-SERVICEPRINCIPALNAME' {
    DependsOn                   = @(
                '[AADApplication]AADApplication-Microsoft365DSC'
    )
    AppId                       = 'SERVICE-PRINCIPAL-ID'
    AppRoleAssignmentRequired   = $False;
    DisplayName                 = 'SERVICEPRINCIPALNAME'
    Ensure                      = 'Present'
    ServicePrincipalType        = 'Application'
    AppRoleAssignedTo           = @(
        MSFT_AADServicePrincipalRoleAssignment {
            PrincipalType = 'User'
            Identity      = 'tenantUser'
        }
        MSFT_AADServicePrincipalRoleAssignment {
            PrincipalType = 'Group'
            Identity      = 'tenantGroup'
        }
    );
    AccountEnabled              = $True;
    ApplicationId               = $ApplicationId
    TenantId                    = $TenantId
    CertificateThumbprint       = $Thumbprint
}
```

### Verbose logs showing the problem

```powershell
VERBOSE: [[AADServicePrincipal]AADServicePrincipal-SERVICEPRINCIPAL::[AzureAD]AzureAD_Configuration] Performing the operation "Update-MgServicePrincipal_UpdateExpanded" on target
 "Call remote 'PATCH /servicePrincipals/{servicePrincipal-id}' operation".
##[error]Cannot bind argument to parameter 'ReferenceObject' because it is null.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.CompareObje 
   ctCommand
    + PSComputerName        : localhost
```

### Environment Information + PowerShell Version

```powershell
OsName               : Microsoft Windows Server 2022 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}




Name                           Value
----                           -----
PSVersion                      5.1.20348.2849
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.20348.2849
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
   * Changed the InsiderRiskTypes property to a string array.
 * AADServicePrincipal
   * Fixes comparison of assigned roles for null values.
+    FIXES [#5717](https://github.com/microsoft/Microsoft365DSC/issues/5717)
 * MISC
   * Changed the CIMInstance logic of various resources to us common logic.
   * Added support for specifying a proxy in Update-M365DSCModule.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,8 @@
 
 * AADConditionalAccessPolicy
   * Changed the InsiderRiskTypes property to a string array.
+* AADServicePrincipal
+  * Fixes comparison of assigned roles for null values.
 * MISC
   * Changed the CIMInstance logic of various resources to us common logic.
   * Added support for specifying a proxy in Update-M365DSCModule.

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_AADServicePrincipal/MSFT_AADServicePrincipal.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_AADServicePrincipal/MSFT_AADServicePrincipal.psm1
@@ -609,6 +609,15 @@ function Set-TargetResource
             [Array]$currentPrincipals = $currentAADServicePrincipal.AppRoleAssignedTo.Identity
             [Array]$desiredPrincipals = $AppRoleAssignedTo.Identity
 
+            if ($null -eq $currentPrincipals)
+            {
+                $currentPrincipals = @()
+            }
+            if ($null -eq $desiredPrincipals)
+            {
+                $desiredPrincipals = @()
+            }
+
             [Array]$differences = Compare-Object -ReferenceObject $currentPrincipals -DifferenceObject $desiredPrincipals
             [Array]$membersToAdd = $differences | Where-Object -FilterScript { $_.SideIndicator -eq '=>' }
             [Array]$membersToRemove = $differences | Where-Object -FilterScript { $_.SideIndicator -eq '<=' }

